Enhance expand command to support reading from stdin and improve code quality

aliamerj opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Consider creating this PR as draft: https://github.com/kubevirt/kubevirt/blob/main/CONTRIBUTING.md#consider-opening-your-pull-request-as-draft
2. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

### What this PR does
Before this PR:

- The virtctl expand command did not support reading from standard input (-).

- Running `./cluster-up/virtctl.sh expand --file -` would result in an error: error reading file open -: no such file or directory.

After this PR:
- The virtctl expand command supports reading from standard input.
- Users can now use the - option to specify standard input for the --file flag.

<!-- (optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*: -->
Fixes #12376

### Why we need it and why it was done in this way
The following tradeoffs were made:
- Supporting standard input for --file flag aligns with common usage patterns in command-line tools, making the tool more flexible and user-friendly.

The following alternatives were considered:
- Keeping the existing behavior without support for standard input. This was deemed less user-friendly and not aligned with common practices.

Links to places where the discussion took place:
- Issue discussion: #12376

Links to places where the discussion took place: <!-- optional: slack, other GH issue, mailinglist, ... -->

### Special notes for your reviewer

<!-- optional -->

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [x] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [x] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

aliamerj commented:
I encountered an issue with my computer when running `./cluster-up/virtctl.sh`  to test my solution locally. Despite this, I have written the code based on my assumptions and unit tests.

Could you please review the changes and let me know if there are any issues? If everything looks good, I will proceed with writing the tests. If there are any errors, please point them out, and I will address them immediately.

Thank you for your understanding and assistance!

aliamerj commented:
\cc @fossedihelm

aliamerj commented:
/cc @fossedihelm

aburdenthehand commented:
/cc @0xFelix

0xFelix commented on pkg/virtctl/vm/expand.go:
```diff
@@ -130,17 +134,27 @@ func expandArgs() cobra.PositionalArgs {
 	}
 
-	err = yml.NewYAMLOrJSONDecoder(bytes.NewReader(readFile), 1024).Decode(&vm)
+	err = yml.NewYAMLOrJSONDecoder(bytes.NewReader(data), 1024).Decode(&vm)
```
nit: You could inline this into the if statement.

0xFelix commented on pkg/virtctl/vm/expand.go:
```diff
@@ -130,17 +134,27 @@ func expandArgs() cobra.PositionalArgs {
+
+	if payload == "-" {
+		data, err = io.ReadAll(os.Stdin)
+		if err != nil {
```
Same, should be possible to inline into if statement.

0xFelix commented on pkg/virtctl/vm/expand.go:
```diff
@@ -130,17 +134,27 @@ func expandArgs() cobra.PositionalArgs {
+		}
+	} else {
+		data, err = os.ReadFile(payload)
+		if err != nil {
```
Same

0xFelix reviewed (commented):
Looks quite good! Is it possible to add a unit test for it?

0xFelix commented:
/test all

aliamerj commented:
@0xFelix 
Thank you for feedback ,
I addressed the inline if statement
now i will try to add some unit test  for the new code

aliamerj commented:
I addded 2 tests
To simulate stdin input in the test, I used an `os.Pipe`. to write test data to one end of the pipe, and read it as stdin from the other end. This approach allows the test to read from stdin as if it were getting user input without requiring any real user interaction.

0xFelix commented on pkg/virtctl/vm/expand_test.go:
```diff
@@ -167,7 +193,28 @@ status:
-			Expect(err).Should(MatchError("error decoding VirtualMachine error converting YAML to JSON: yaml: mapping values are not allowed in this context"))
+			Expect(err).Should(MatchError("error decoding VirtualMachine: error converting YAML to JSON: yaml: mapping values are not allowed in this context"))
+		})
+		It("should fail when called with invalid yaml in stdin input", func() {
```
nit: Add newline above

0xFelix commented on pkg/virtctl/vm/expand_test.go:
If you want you could deduplicate tests in this file by using `DescribeTable`. This would allow for a great clean up by removing unnecessary deduplications of code.

0xFelix commented on pkg/virtctl/vm/expand_test.go:
```diff
@@ -145,6 +145,32 @@ status:
+		It("should succeed when called with valid stdin input", func() {
+			// Mock stdin
+			oldStdin := os.Stdin
+			defer func() { os.Stdin = oldStdin }()
```
Can you use ginkgo's `DeferCleanup` here?

0xFelix commented on pkg/virtctl/vm/expand_test.go:
```diff
@@ -145,6 +145,32 @@ status:
+			expandSpecInterface.EXPECT().ForVirtualMachine(vm).Return(vm, nil).Times(1)
+
+			cmd := clientcmd.NewRepeatableVirtctlCommand("expand", fileInput, "-")
+			err = cmd()
+
+			Expect(err).ToNot(HaveOccurred())
```
```suggestion
			Expect(cmd()).To(Succeed())
```

0xFelix commented on pkg/virtctl/vm/expand_test.go:
Also, could you add the new `stdin` tests outside of the `Context("with file input", ...`?

aliamerj commented:
@0xFelix 
How about using  Context for stdin input, it looks much cleaner and simple
i tried also to put the  stdin mock  in the before Each but it did not work
```go
oldStdin := os.Stdin
			r, w, err := os.Pipe()
			os.Stdin = r
			DeferCleanup(func() {
				os.Stdin = oldStdin
				r.Close()
				w.Close()
			})
			Expect(err).ToNot(HaveOccurred())
			os.Stdin = r
```
what do yout think ?

0xFelix commented:
I like it. What did not work when putting it in BeforeEach? Would be great if you could make it work.

aliamerj commented:
@0xFelix 
I figured out what my mistake was. I was calling this function in BeforeEach:
```go
go func() {
defer w.Close()
_, _ = w.Write([]byte(data))
}()
```
This caused the write end of the pipe to close before the test started. I solved it by making the function below and call it in top of  the tests that required the stdin:
``go
	writeToStdin := func(data string) {
			go func() {
				defer w.Close()
				_, _ = w.Write([]byte(data))
			}()
		}
```go

0xFelix commented on pkg/virtctl/vm/expand_test.go:
```diff
@@ -167,7 +167,56 @@ status:
+			w        *os.File
+			err      error
+		)
+		writeToStdin := func(data string) {
```
nit: add newline above

0xFelix commented on pkg/virtctl/vm/expand_test.go:
```diff
@@ -167,7 +167,56 @@ status:
+		It("should succeed when called with valid stdin input", func() {
+			writeToStdin(vmSpec)
+
+			Expect(yaml.Unmarshal([]byte(vmSpec), vm)).To(Succeed())
```
I don't think this assertion is required here?

0xFelix commented on pkg/virtctl/vm/expand_test.go:
```diff
@@ -167,7 +167,56 @@ status:
+			cmd := clientcmd.NewRepeatableVirtctlCommand("expand", fileInput, "-")
+			err = cmd()
+
+			Expect(err).To(HaveOccurred())
```
You can drop this assertion, it is covered by the line below (`MatchError`).

0xFelix commented on pkg/virtctl/vm/expand_test.go:
```diff
@@ -167,7 +167,56 @@ status:
+			writeToStdin(invalidYaml)
+
+			cmd := clientcmd.NewRepeatableVirtctlCommand("expand", fileInput, "-")
+			err = cmd()
+
+			Expect(err).To(HaveOccurred())
+			Expect(err).Should(MatchError("error decoding VirtualMachine: error converting YAML to JSON: yaml: mapping values are not allowed in this context"))
```
```suggestion
			Expect(cmd()).To(MatchError("error decoding VirtualMachine: error converting YAML to JSON: yaml: mapping values are not allowed in this context"))
```

0xFelix commented on pkg/virtctl/vm/expand_test.go:
```diff
@@ -167,7 +167,56 @@ status:
+		})
+
+		It("should succeed when called with valid stdin input", func() {
+			writeToStdin(vmSpec)
```
nit: Put the `EXPECT()` calls first and move this down above `NewRepeatableVirtctlCommand`

0xFelix reviewed (commented):
It looks quite good, some stylistic fixes and we can merge it.

0xFelix commented on pkg/virtctl/vm/expand_test.go:
```diff
@@ -167,7 +167,56 @@ status:
+		})
+
+		It("should succeed when called with valid stdin input", func() {
+			yaml.Unmarshal([]byte(vmSpec), vm)
```
Ah, now I understood why the assertion was needed. Sorry, that was a misunderstanding on my part. Please re-add the `Expect`.

0xFelix commented on pkg/virtctl/vm/expand_test.go:
```diff
@@ -167,7 +167,56 @@ status:
 			err := cmd()
 
 			Expect(err).To(HaveOccurred())
-			Expect(err).Should(MatchError("error decoding VirtualMachine error converting YAML to JSON: yaml: mapping values are not allowed in this context"))
+			Expect(err).Should(MatchError("error decoding VirtualMachine: error converting YAML to JSON: yaml: mapping values are not allowed in this context"))
```
nit: You can drop `Expect(err).To(HaveOccurred())`

0xFelix reviewed (commented):
Two comments. To me it looks good otherwise.

/approve

aliamerj commented:
@0xFelix 
i made the change

0xFelix commented:
Thanks

@fossedihelm PTAL

fossedihelm commented on pkg/virtctl/vm/expand_test.go:
```diff
@@ -164,10 +164,57 @@ status:
+
+		writeToStdin := func(data string) {
+			go func() {
+				defer w.Close()
```
```
// Close will return an error if it has already been called.
```
Doesn't this `w.Close()` conflict with the one performed in the `DeferCleanup`?

fossedihelm commented:
/test pull-kubevirt-unit-test

fossedihelm commented:
/test all

fossedihelm commented:
@aliamerj Failure is related: https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/12414/pull-kubevirt-e2e-k8s-1.30-sig-compute/1818515080656982016
Can you adjust the e2e test error? Thank you!

aliamerj commented:
@fossedihelm 
I found the issue: it was a mismatch in error message formatting. I had updated the function that reads from a file or stdin and mistakenly changed the error message format from %+w to %w. 
I've now corrected this. 
Please rerun the e2e tests, and I'll fix any further issues if they arise.

fossedihelm commented:
Thank you!
/test all

fossedihelm commented:
/lgtm

fossedihelm commented:
/lgtm
